repose-aggregator/components/filters/header-id-mapping/src/main/scala/org/openrepose/filters/headeridmapping/HeaderIdMappingFilter.scala

```diff
@@ -0,0 +1,97 @@
+/*
+ * _=_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=
+ * Repose
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Copyright (C) 2010 - 2016 Rackspace US, Inc.
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+
+package org.openrepose.filters.headeridmapping
+
+import java.net.URL
+import javax.inject.{Inject, Named}
+import javax.servlet._
+import javax.servlet.http.HttpServletRequest
+
+import com.typesafe.scalalogging.slf4j.LazyLogging
+import org.openrepose.commons.config.manager.UpdateListener
+import org.openrepose.commons.utils.http.PowerApiHeader
+import org.openrepose.commons.utils.servlet.http.HttpServletRequestWrapper
+import org.openrepose.core.filter.FilterConfigHelper
+import org.openrepose.core.services.config.ConfigurationService
+import org.openrepose.filters.headeridmapping.config.{HttpHeader, HeaderIdMappingConfig}
+
+import scala.collection.JavaConverters._
+
+@Named
+class HeaderIdMappingFilter @Inject()(configurationService: ConfigurationService)
+  extends Filter with UpdateListener[HeaderIdMappingConfig] with LazyLogging {
+
+  import HeaderIdMappingFilter._
+
+  private var configurationFile: String = DEFAULT_CONFIG
+  private var initialized = false
+  private var configuredHeaders: Iterable[HttpHeader] = _
+
+  override def init(filterConfig: FilterConfig): Unit = {
+    logger.trace("Header Identity Mapping filter initializing...")
+    configurationFile = new FilterConfigHelper(filterConfig).getFilterConfig(DEFAULT_CONFIG)
+
+    logger.info(s"Initializing Header Identity Mapping filter using config $configurationFile")
+    val xsdUrl: URL = getClass.getResource(SCHEMA_FILE_NAME)
+    configurationService.subscribeTo(filterConfig.getFilterName, configurationFile, xsdUrl, this, classOf[HeaderIdMappingConfig])
+
+    logger.trace("Header Identity Mapping filter initialized.")
+  }
+
+  override def doFilter(servletRequest: ServletRequest, servletResponse: ServletResponse, filterChain: FilterChain): Unit = {
+    val wrappedRequest = new HttpServletRequestWrapper(servletRequest.asInstanceOf[HttpServletRequest])
```

Comment:
That sounds like a good idea. Even before it could vary based on the implementation of the HandlerFactory, which wasn't great.

Design wise, only the UpdateListener will know when it has been passed a configuration. Without changing any interfaces, that means that we probably don't know if a filter has read the configuration until we try sending it a request. Which means that logic being invoked by every filter, since we can't change the Filter interface.

One alternative would be, perhaps, to amend the ConfigurationService to intelligently handle configuration processing. Then we could have some centralized code (say in the PowerFilter) that asks the ConfigurationService is all configurations have been processed. If not, reject with a 503, if so, send the request through the chain.

In the latter case, I would suggest putting that work in a new story. Should we still, then, add the logic to each filter for the time being?

Paging @adrianjgeorge.